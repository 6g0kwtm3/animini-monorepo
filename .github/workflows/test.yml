name: Test

on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master

permissions:
  contents: read

jobs:
  test:
    name: Test E2E
    runs-on: ubuntu-latest
    container:
      image: mcr.microsoft.com/playwright:v1.43.1-jammy

    steps:
      - uses: actions/checkout@v3
      - uses: pnpm/action-setup@v2
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: "pnpm"

      - name: Install dependencies ðŸ“¦
        run: pnpm i --frozen-lockfile

      - name: Await CF Pages
        uses: WalshyDev/cf-pages-await@v1
        id: cf-pages
        with:
          apiToken: ${{ secrets.CF_API_TOKEN  }}
          accountId: 'e9b0348e83d33e46fbb82c3271317d70'
          project: 'animini-pwa'

      - name: Test E2E
        run: pnpm run test
        env:
          ANILIST_TEST_TOKEN: ${{ secrets.ANILIST_TEST_TOKEN }}
          BASE_URL: ${{ steps.cf-pages.outputs.url }}
          HOME: /root

  typecheck:
    runs-on: ubuntu-latest

    name: Typecheck
    steps:
      - uses: actions/checkout@v3
      - uses: pnpm/action-setup@v2
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: "pnpm"

      - name: Install dependencies ðŸ“¦
        run: pnpm i --frozen-lockfile

      - run: pnpm run paraglide

      - name: Typecheck
        run: pnpm run check

  format:
    runs-on: ubuntu-latest
    name: Lint & Format
    permissions:
      contents: write

    steps:
      - uses: actions/checkout@v3
        with:
          token: ${{ secrets.PAT }}
      - uses: pnpm/action-setup@v2
      - uses: actions/setup-node@v3
        with:
          node-version: 20
          cache: "pnpm"

      - name: Install dependencies ðŸ“¦
        run: pnpm i --frozen-lockfile

      - name: Lint
        run: pnpm run lint --fix

      - name: Format
        run: pnpm run format

      - name: Commit changes
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: "refactor: run format"
