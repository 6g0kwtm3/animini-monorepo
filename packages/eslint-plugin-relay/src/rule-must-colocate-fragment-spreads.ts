/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of thttps://github.com/relayjs/eslint-plugin-relay.
 *
 * This rule lints for non-colocated fragment spreads within queries or
 * fragments. In other words, situations where a fragment is spread in module A,
 * but the module (B) that defines that fragment is not imported by module A.
 * It does not lint subscriptions or mutations. This catches:
 *
 * - The anti-pattern of spreading a fragment in a parent module, then passing
 * that data down to a child module, or jamming it all in context. This defeats
 * the purpose of Relay. From the
 * [Relay docs](https://relay.dev/docs/next) â€“ "[Relay]
 * allows components to specify what data they need and the Relay framework
 * provides the data. This makes the data needs of inner components opaque and
 * allows composition of those needs."
 * - Instances where fragment spreads are unused, which results in overfetching.
 *
 * ## When the fragment is unused
 * The easiest way to tell if a fragment is unused is to remove the line
 * containing the lint error, run Relay compiler, then Flow. If there are no
 * type errors, then the fragment was possibly unused. You should still test
 * your functionality to see that it's working as expected.
 *
 * ## When the fragment is being passed to a child component
 * If you received Relay or Flow errors after attempting to remove the fragment,
 * then it's very likely that you're passing that data down the tree. Our
 * recommendation is to have components specify the data they need. In the below
 * example, this is an anti-pattern because Component B's data requirements are
 * no longer opaque. Component B should not be fetching data on Component C's
 * behalf.
 *
 * function ComponentA(props) {
 *   const data = useFragment(graphql`
 *     fragment ComponentA_fragment on User {
 *       foo
 *       bar
 *       some_field {
 *         ...ComponentC_fragment
 *       }
 *     }
 *   `);
 *   return (
 *     <div>
 *       {data.foo} {data.baz}
 *       <ComponentB text="Hello" data={data.some_field} />
 *     </div>
 *   );
 * }
 *
 * To address this, refactor Component C to fetch the data it needs. You'll need
 * to update the intermediate components by amending, or adding a fragment to
 * each intermediate component between ComponentA and ComponentC.
 */

"use strict"

import type { Rule } from "eslint"
import {
	BREAK,
	Kind,
	OperationTypeNode,
	visit,
	type DocumentNode,
	type FragmentSpreadNode,
} from "graphql"
import {
	getGraphQLAST,
	getLoc,
	getModuleName,
	hasPrecedingEslintDisableComment,
	isGraphQLTemplate,
	type GraphqlTemplateExpression,
} from "./utils"

const ESLINT_DISABLE_COMMENT =
	" eslint-disable-next-line eslint-plugin-relay/must-colocate-fragment-spreads"

function isReadonlyArray(value: unknown): value is readonly unknown[] {
	return Array.isArray(value)
}

function getGraphQLFragmentSpreads(graphQLAst: DocumentNode) {
	const fragmentSpreads: Record<string, FragmentSpreadNode> = {}
	visit(graphQLAst, {
		FragmentSpread(node, key, parent, path, ancestors) {
			for (const ancestorNode of ancestors) {
				if (isReadonlyArray(ancestorNode)) {
					continue
				}
				if (
					ancestorNode.kind === Kind.OPERATION_DEFINITION
					&& (ancestorNode.operation === OperationTypeNode.MUTATION
						|| ancestorNode.operation === OperationTypeNode.SUBSCRIPTION)
				) {
					return
				}
			}
			for (const directiveNode of node.directives ?? []) {
				if (directiveNode.name.value === "module") {
					return
				}
				if (directiveNode.name.value === "relay") {
					for (const argumentNode of directiveNode.arguments ?? []) {
						if (
							argumentNode.name.value === "mask"
							&& "value" in argumentNode.value
							&& argumentNode.value.value === false
						) {
							return
						}
					}
				}
			}
			if (hasPrecedingEslintDisableComment(node, ESLINT_DISABLE_COMMENT)) {
				return
			}
			fragmentSpreads[node.name.value] = node
		},
	})
	return fragmentSpreads
}

function getGraphQLFragmentDefinitionName(
	graphQLAst: DocumentNode
): string | null {
	let name: string | null = null
	visit(graphQLAst, {
		FragmentDefinition(node) {
			name = node.name.value
			return BREAK
		},
	})
	return name
}

export const rule: Rule.RuleModule = {
	meta: {
		docs: {},
		schema: [],
		messages: {
			"must-colocate-fragment-spreads":
				`This spreads the fragment \`{{ fragment }}\` but `
				+ "this module does not use it directly.",
		},
	},
	create(context) {
		const foundImportedModules: string[] = []
		const graphqlLiterals: {
			node: GraphqlTemplateExpression
			graphQLAst: DocumentNode
		}[] = []

		return {
			"Program:exit"(_node) {
				const fragmentsInTheSameModule = new Set<string>()
				graphqlLiterals.forEach(({ graphQLAst }) => {
					const fragmentName = getGraphQLFragmentDefinitionName(graphQLAst)
					if (fragmentName) {
						fragmentsInTheSameModule.add(fragmentName)
					}
				})
				graphqlLiterals.forEach(({ node, graphQLAst }) => {
					const queriedFragments = getGraphQLFragmentSpreads(graphQLAst)
					for (const fragment in queriedFragments) {
						const matchedModuleName = foundImportedModules.find((name) =>
							fragment.startsWith(name)
						)
						if (
							queriedFragments[fragment]
							&& !matchedModuleName
							&& !fragmentsInTheSameModule.has(fragment)
						) {
							context.report({
								node,
								messageId: "must-colocate-fragment-spreads",
								loc: getLoc(context, node, queriedFragments[fragment]),
								data: { fragment },
							})
						}
					}
				})
			},

			ImportDeclaration(node) {
				if (
					"importKind" in node
					&& node.importKind === "value"
					&& typeof node.source.value === "string"
				) {
					foundImportedModules.push(getModuleName(node.source.value))
				}
			},

			ImportExpression(node) {
				if (
					node.source.type === "Literal"
					&& typeof node.source.value === "string"
				) {
					// Allow dynamic imports like import(`test/${fileName}`); and (path) => import(path);
					// These would have node.source.value undefined
					foundImportedModules.push(getModuleName(node.source.value))
				}
			},

			CallExpression(node) {
				if ("name" in node.callee && node.callee.name !== "require") {
					return
				}
				const [source] = node.arguments
				if (
					source
					&& source.type === "Literal"
					&& typeof source.value === "string"
				) {
					foundImportedModules.push(getModuleName(source.value))
				}
			},

			TaggedTemplateExpression(node) {
				if (isGraphQLTemplate(node)) {
					const graphQLAst = getGraphQLAST(node)
					if (!graphQLAst) {
						// ignore nodes with syntax errors, they're handled by rule-graphql-syntax
						return
					}
					graphqlLiterals.push({ node, graphQLAst })
				}
			},
		}
	},
}
