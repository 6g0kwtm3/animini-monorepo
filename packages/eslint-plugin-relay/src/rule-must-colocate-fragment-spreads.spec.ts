import * as typescriptParser from "@typescript-eslint/parser"
import { createRuleTester } from "eslint-vitest-rule-tester"
import { describe } from "node:test"
import { expect, test } from "vitest"
import { rule } from "./rule-must-colocate-fragment-spreads"

void describe("rule-must-colocate-fragment-spreads", () => {
	const { valid, invalid } = createRuleTester({
		rule,
		languageOptions: { ecmaVersion: 6, parser: typescriptParser },
	})

	test("allows fragment spread when component is imported", async () => {
		await valid(
			`
      import { Component } from '../shared/component.js';
      graphql\`fragment foo on Page {
        ...component_fragment
      }\`;
      `
		)
	})

	test("allows fragment spread when component is required", async () => {
		await valid(
			`
      const Component = require('../shared/component.js');
      graphql\`fragment foo on Page {
        ...component_fragment
      }\`;
      `
		)
	})

	test("allows fragment spread when component is dynamically imported", async () => {
		await valid(
			`
      const Component = import('../shared/component.js');
      graphql\`fragment foo on Page {
        ...component_fragment
      }\`;
      `
		)
	})

	test("allows fragment spread with nested component path", async () => {
		await valid(
			`
      import { Component } from './nested/componentModule.js';
      graphql\`fragment foo on Page {
        ...componentModule_fragment
      }\`;
      `
		)
	})

	test("allows fragment spread with component module naming convention", async () => {
		await valid(
			`
      import { Component } from './component-module.js';
      graphql\`fragment foo on Page {
        ...componentModuleFragment
      }\`;
      `
		)
	})

	test("allows fragment spread in query", async () => {
		await valid(
			`
      import { Component } from './component-module.js';
      graphql\`query Root {
        ...componentModuleFragment
      }\`;
      `
		)
	})

	test("allows local fragment spreads within same file", async () => {
		await valid(
			`
      graphql\`fragment foo1 on Page {
        name
      }\`;
      graphql\`fragment foo2 on Page {
        ...foo1
      }\`;
      `
		)
	})

	test("allows fragment spreads in mutations", async () => {
		await valid(
			`
      graphql\`mutation {
        page_unlike(data: $input) {
          ...component_fragment
          ...componentFragment
          ...component
        }
      }\`
      `
		)
	})

	test("allows fragment spread with relay mask directive", async () => {
		await valid(
			`
      graphql\`fragment foo on Page { ...Fragment @relay(mask: false) }\`;
      `
		)
	})

	test("allows fragment spread with relay module directive", async () => {
		await valid(
			`
      graphql\`fragment foo on Page { ...Fragment @module(name: "ComponentName.react") }\`;
      `
		)
	})

	test("allows dynamic imports with template literal paths", async () => {
		await valid(
			"\
      const getOperation = (reference) => {\
        return import(`./src/__generated__/${reference}`);\
      };\
      "
		)
	})

	test("allows dynamic imports with variable paths", async () => {
		await valid(
			"\
      const getOperation = (reference) => {\
        return import(reference);\
      };\
      "
		)
	})

	test("allows fragment spread with eslint disable comment", async () => {
		await valid(
			`
      graphql\`fragment foo on Page {
        # eslint-disable-next-line eslint-plugin-relay/must-colocate-fragment-spreads
        ...unused1
      }\`;
      `
		)
	})

	test("reports error for unused fragment spread in fragment", async () => {
		const { result } = await invalid({
			code: `
        graphql\`fragment foo on Page { ...unused1 }\`;
        `,
			errors: ["must-colocate-fragment-spreads"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports errors for multiple unused fragment spreads", async () => {
		const { result } = await invalid({
			code: `
        graphql\`fragment Test on Page { ...unused1, ...unused2 }\`;
        `,
			errors: [
				"must-colocate-fragment-spreads",
				"must-colocate-fragment-spreads",
			],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for unused fragment spread in query", async () => {
		const { result } = await invalid({
			code: `
        graphql\`query Root { ...unused1 }\`;
        `,
			errors: ["must-colocate-fragment-spreads"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for unused fragment spread even with used fragments", async () => {
		const { result } = await invalid({
			code: `
        import { Component } from './used1.js';
        graphql\`fragment foo on Page { ...used1 ...unused1 }\`;
        `,
			errors: ["must-colocate-fragment-spreads"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for unused fragment spread with relay mask directive", async () => {
		const { result } = await invalid({
			code: `
        graphql\`fragment foo on Page { ...unused1 @relay(mask: true) }\`;
        `,
			errors: ["must-colocate-fragment-spreads"],
		})
		expect(result.output).toMatchSnapshot()
	})

	test("reports error for unused component fragment", async () => {
		const { result } = await invalid({
			code: `
        import type { MyType } from '../shared/component.js';
        graphql\`fragment foo on Page {
          ...component_fragment
        }\`;
        `,
			errors: ["must-colocate-fragment-spreads"],
		})
		expect(result.output).toMatchSnapshot()
	})
})
