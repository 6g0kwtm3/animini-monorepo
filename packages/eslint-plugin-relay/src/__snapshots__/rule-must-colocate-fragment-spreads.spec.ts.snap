// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reports error for unused component fragment 1`] = `
"
        import type { MyType } from '../shared/component.js';
        graphql\`fragment foo on Page {
          ...component_fragment
        }\`;
        "
`;

exports[`reports error for unused fragment spread even with used fragments 1`] = `
"
        import { Component } from './used1.js';
        graphql\`fragment foo on Page { ...used1 ...unused1 }\`;
        "
`;

exports[`reports error for unused fragment spread in fragment 1`] = `
"
        graphql\`fragment foo on Page { ...unused1 }\`;
        "
`;

exports[`reports error for unused fragment spread in query 1`] = `
"
        graphql\`query Root { ...unused1 }\`;
        "
`;

exports[`reports error for unused fragment spread with relay mask directive 1`] = `
"
        graphql\`fragment foo on Page { ...unused1 @relay(mask: true) }\`;
        "
`;

exports[`reports errors for multiple unused fragment spreads 1`] = `
"
        graphql\`fragment Test on Page { ...unused1, ...unused2 }\`;
        "
`;
